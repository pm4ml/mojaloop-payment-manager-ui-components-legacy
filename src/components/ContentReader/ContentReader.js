import 'highlight.js/styles/googlecode.css';
import './ContentReader.scss';

import hljs from 'highlight.js/lib/highlight';
import json from 'highlight.js/lib/languages/json';
import shell from 'highlight.js/lib/languages/shell';
import xml from 'highlight.js/lib/languages/xml';
import React, { PureComponent } from 'react';
import vkbeautify from 'vkbeautify';

import ScrollBox from '../ScrollBox';

hljs.registerLanguage('xml', xml);
hljs.registerLanguage('json', json);
hljs.registerLanguage('shell', shell);

function restrictKeysToSelectAndCopy(evt) {
  const { ctrlKey, metaKey, keyCode } = evt;
  const isCopyOrSelectKey = keyCode === 65 || keyCode === 67;
  const isCopyOrSelectModifier = ctrlKey || metaKey;
  if (!isCopyOrSelectKey || !isCopyOrSelectModifier) {
    evt.preventDefault();
  }
}

class ContentReader extends PureComponent {
  static parse(source = '') {
    let content = source;
    let lineNumbers = null;
    let error = false;

    try {
      content = vkbeautify.json(source, 2);
    } catch (jsonErr) {
      try {
        content = vkbeautify.xml(source, 2);
      } catch (xmlErr) {
        error = true;
      }
    }

    if (!error) {
      const lines = content.split(/(?:\r\n|\r|\n)/);
      lineNumbers = lines.map((_, index) => (
        <div className="content-reader__lines__line-n" key={index.toString()}>
          {index + 1}
        </div>
      ));
    }

    return { content, lineNumbers, error };
  }
  componentDidMount() {
    hljs.highlightBlock(this.code);
  }
  componentDidUpdate() {
    // remove className preventing re-highlighting
    this.code.className = '';
    hljs.highlightBlock(this.code);
  }

  render() {
    const { content, lineNumbers, error } = ContentReader.parse(this.props.data);
    if (error) {
      return <div>Unable to read the data</div>;
    }

    return (
      <div className="content-reader">
        <div className="content-reader__layout-v" style={this.props.style}>
          <ScrollBox flex>
            <div className="content-reader__layout-h">
              <div className="content-reader__lines">
                <pre>
                  <code>{lineNumbers}</code>
                </pre>
              </div>
              <div
                suppressContentEditableWarning
                contentEditable
                spellCheck="false"
                className="content-reader__content"
                role="presentation"
                onKeyDown={restrictKeysToSelectAndCopy}
              >
                <pre>
                  <code
                    ref={code => {
                      this.code = code;
                    }}
                  >
                    {content}
                  </code>
                </pre>
              </div>
            </div>
          </ScrollBox>
        </div>
      </div>
    );
  }
}

export default ContentReader;
