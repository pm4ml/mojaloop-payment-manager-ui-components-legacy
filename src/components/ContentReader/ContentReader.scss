.content-reader {
  display: flex;
  flex: 1;
  flex-direction: column;
  max-width: 100%;
  border: 2px solid #eee;
  min-height: 0;

  pre {
    margin: 0;
    padding: 0;
    word-wrap: normal;
  }

  code {
    font-family: monospace;
    color: #444;
    line-height: 16px;
  }

  span {
    display: inline-block;
  }

  .hljs {
    padding: 0px;
    background: none;
    overflow: visible;
  }
}
.content-reader__layout-v {
  display: flex;
  flex: 2;
  max-height: 100%;
}

.content-reader__layout-h {
  display: flex;
  flex: 2;
  align-items: stretch;
  flex-direction: row;
  min-height: min-content;
  max-width: 100%;
}

.content-reader__lines {
  flex: 0 0 50px;
  padding: 2px 4px;
  background: #f8f8f8;
  border-right: 2px solid #f4f4f4;
  text-align: right;
  min-height: min-content;

  code {
    font-weight: bold;
    color: #aaa;
  }
}
.content-reader__lines__line-n {
  height: 16px;
}

.content-reader__content {
  flex: 1 1 auto;
  overflow-x: auto;
  min-height: min-content;
  display: flex;
  outline: none;
  pre {
    min-height: min-content;
  }
}
