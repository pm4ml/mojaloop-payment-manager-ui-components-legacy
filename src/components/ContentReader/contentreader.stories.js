import React from 'react';

import Row from '../Row';
import ContentReader from './ContentReader';

export default {
  title: 'ContentReader',
  component: ContentReader,
};

const json = `{
  "Interchange": {
    "DataElementSeparator": "*",
    "SegmentSeparat123or1": "~%WSP*; %NL;%WSP*;",
    "SegmentSeparato123r": "~%WSP*; %NL;%WSP*;",
    "SegmentSepar1ator": "~%WSP*; %NL;%WSP*;",
    "Segmen1tSepara2tor": "~%WSP*; %NL;%WSP*;",
    "Segment213Separat2or": "~%WSP*; %NL;%WSP*;",
    "SegmentSe23parato4r": "~%WSP*; %NL;%WSP*;",
    "aSegmentSepa23rato6r": "~%WSP*; %NL;%WSP*;",
    "SegmentSeparasdasa231to7r": "~%WSP*; %NL;%WSP*;",
    "ISA": {
      "ISA15": "Test",
      "ISA16": ">"
    }
  }
}`;

const xml = `<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">
  <xs:element name="items" type="ItemsType"/>
  <xs:complexType name="ItemsType">
    <xs:choice minOccurs="0" maxOccurs="unbounded">
      <xs:element name="shirt" type="ProductType"/>
      <xs:element name="hat" type="ProductType"/>
      <xs:element name="umbrella" type="ProductType"/>
    </xs:choice>
  </xs:complexType>
  <!-- Element only content -->
  <xs:complexType name="ProductType">
    <xs:sequence>
      <xs:element name="number" type="xs:integer"/>
      <xs:element name="name" type="xs:string"/>
      <xs:choice minOccurs="0" maxOccurs="unbounded">
        <xs:element name="size" type="SizeType"/>
        <xs:element name="color" type="ColorType"/>
        <xs:element name="description" type="DescriptionType"/>
      </xs:choice>
    </xs:sequence>
    <xs:attribute  name="effDate" type="xs:date"
                    default="1900-01-01"/>
    <xs:anyAttribute namespace="##other" processContents="lax"/>
  </xs:complexType>
  <!-- Simple content -->
  <xs:complexType name="SizeType">
    <xs:simpleContent>
      <xs:extension base="xs:integer">
        <xs:attribute name="system" type="xs:token"/>
      </xs:extension>
    </xs:simpleContent>
  </xs:complexType>
  <!-- Empty content -->
  <xs:complexType name="ColorType">
    <xs:attribute name="value" type="xs:string"/>
  </xs:complexType>
  <!-- Mixed content -->
  <xs:complexType name="DescriptionType" mixed="true">
    <xs:sequence>
      <xs:any namespace="http://www.w3.org/1999/xhtml"
      minOccurs="0" maxOccurs="unbounded"
      processContents="skip"/>
    </xs:sequence>
  </xs:complexType>
</xs:schema>
`;

export const ParsingJSONContent = () => (
  <Row style={{ maxHeight: '100%' }}>
    <ContentReader data={json} />
  </Row>
);

export const ParsingXMLContent = () => (
  <Row style={{ maxHeight: '100%' }}>
    <ContentReader data={xml} />
  </Row>
);
