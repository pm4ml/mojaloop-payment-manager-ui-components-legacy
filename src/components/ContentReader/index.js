import ContentReader from './ContentReader';

export default ContentReader;
