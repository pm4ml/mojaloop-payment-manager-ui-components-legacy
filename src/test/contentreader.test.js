import { mount, shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import React from 'react';

import ContentReader from '../components/ContentReader';

// Shallow render, no sub-components ( still available children html)
const jsonObject = [{ name: 'test-name', value: 'test-value' }];
const jsonString = JSON.stringify(jsonObject);
const xmlString = `<note><to>Tove</to><from>Jani</from><xyz>test</xyz><body>test</body></note>`;
const shellString = `$ echo $EDITOR
                      vim
                      $ git checkout master
                      Switched to branch 'master'
                      Your branch is up-to-date with 'origin/master'.
                      $ git push
                      Everything up-to-date
                      `;

it('renders the component', () => {
  const wrapper = mount(<ContentReader data={jsonString} />);
  expect(wrapper.find('.content-reader').exists()).toBe(true);
});

it('generates the correct line count', () => {
  const wrapper = mount(<ContentReader data={jsonString} />);
  expect(wrapper.find('.content-reader__lines__line-n')).toHaveLength(6);
});

it('applies the correct json className', async () => {
  const wrapper = mount(<ContentReader data={jsonString} />);
  const classNames = wrapper
    .find('.content-reader__content')
    .find('code')
    .getDOMNode()
    .className.split(' ');

  expect(classNames.includes('hljs')).toBe(true);
  expect(classNames.includes('json')).toBe(true);
});

it('applies the correct xml className', async () => {
  const wrapper = mount(<ContentReader data={xmlString} />);
  const classNames = wrapper
    .find('.content-reader__content')
    .find('code')
    .getDOMNode()
    .className.split(' ');

  expect(classNames.includes('hljs')).toBe(true);
  expect(classNames.includes('xml')).toBe(true);
});

it('applies the correct shell className', async () => {
  const wrapper = mount(<ContentReader data={shellString} />);
  const classNames = wrapper
    .find('.content-reader__content')
    .find('code')
    .getDOMNode()
    .className.split(' ');

  expect(classNames.includes('hljs')).toBe(true);
  expect(classNames.includes('shell')).toBe(true);
});

it('renders the ContentReader correctly when multiple props are set', () => {
  const wrapper = mount(<ContentReader data={shellString} />);
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});
