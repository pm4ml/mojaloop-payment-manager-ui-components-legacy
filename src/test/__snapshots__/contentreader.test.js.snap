// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the ContentReader correctly when multiple props are set 1`] = `
<ContentReader
  data="$ echo $EDITOR
                      vim
                      $ git checkout master
                      Switched to branch 'master'
                      Your branch is up-to-date with 'origin/master'.
                      $ git push
                      Everything up-to-date
                      "
>
  <div
    className="content-reader"
  >
    <div
      className="content-reader__layout-v"
    >
      <ScrollBox
        flex={true}
        handleStyle={Object {}}
        showTrack={false}
        style={Object {}}
        trackStyle={Object {}}
      >
        <ResizeDetector
          handleHeight={true}
          handleWidth={true}
          nodeType="div"
          onResize={[Function]}
          querySelector={null}
          refreshRate={1000}
          skipOnMount={false}
          targetDomEl={null}
          targetRef={
            Object {
              "current": <div
                class="element el-scrollbox__wrapper"
              >
                <div
                  class="el-scrollbox__content-box el-scrollbox__content-box--flexible"
                >
                  <div
                    class="el-scrollbox__content el-scrollbox__content--flexible"
                    style="min-width: NaNpx; max-width: NaNpx;"
                  >
                    <div
                      class="content-reader__layout-h"
                    >
                      <div
                        class="content-reader__lines"
                      >
                        <pre>
                          <code>
                            <div
                              class="content-reader__lines__line-n"
                            >
                              1
                            </div>
                            <div
                              class="content-reader__lines__line-n"
                            >
                              2
                            </div>
                            <div
                              class="content-reader__lines__line-n"
                            >
                              3
                            </div>
                            <div
                              class="content-reader__lines__line-n"
                            >
                              4
                            </div>
                            <div
                              class="content-reader__lines__line-n"
                            >
                              5
                            </div>
                            <div
                              class="content-reader__lines__line-n"
                            >
                              6
                            </div>
                            <div
                              class="content-reader__lines__line-n"
                            >
                              7
                            </div>
                            <div
                              class="content-reader__lines__line-n"
                            >
                              8
                            </div>
                          </code>
                        </pre>
                      </div>
                      <div
                        class="content-reader__content"
                        contenteditable="true"
                        role="presentation"
                        spellcheck="false"
                      >
                        <pre>
                          <code
                            class="hljs shell"
                          >
                            <span
                              class="hljs-meta"
                            >
                              $
                            </span>
                             echo $EDITOR
                      vim
                      $ git checkout master
                      Switched to branch 'master'
                      Your branch is up-to-date with 'origin/master'.
                      $ git push
                      Everything up-to-date
                      
                          </code>
                        </pre>
                      </div>
                    </div>
                  </div>
                </div>
              </div>,
            }
          }
        >
          <div
            className="element el-scrollbox__wrapper"
            style={Object {}}
          >
            <div
              className="el-scrollbox__content-box el-scrollbox__content-box--flexible"
            >
              <ResizeDetector
                handleHeight={true}
                handleWidth={true}
                nodeType="div"
                onResize={[Function]}
                querySelector={null}
                refreshRate={1000}
                skipOnMount={false}
                targetDomEl={null}
                targetRef={
                  Object {
                    "current": <div
                      class="el-scrollbox__content el-scrollbox__content--flexible"
                      style="min-width: NaNpx; max-width: NaNpx;"
                    >
                      <div
                        class="content-reader__layout-h"
                      >
                        <div
                          class="content-reader__lines"
                        >
                          <pre>
                            <code>
                              <div
                                class="content-reader__lines__line-n"
                              >
                                1
                              </div>
                              <div
                                class="content-reader__lines__line-n"
                              >
                                2
                              </div>
                              <div
                                class="content-reader__lines__line-n"
                              >
                                3
                              </div>
                              <div
                                class="content-reader__lines__line-n"
                              >
                                4
                              </div>
                              <div
                                class="content-reader__lines__line-n"
                              >
                                5
                              </div>
                              <div
                                class="content-reader__lines__line-n"
                              >
                                6
                              </div>
                              <div
                                class="content-reader__lines__line-n"
                              >
                                7
                              </div>
                              <div
                                class="content-reader__lines__line-n"
                              >
                                8
                              </div>
                            </code>
                          </pre>
                        </div>
                        <div
                          class="content-reader__content"
                          contenteditable="true"
                          role="presentation"
                          spellcheck="false"
                        >
                          <pre>
                            <code
                              class="hljs shell"
                            >
                              <span
                                class="hljs-meta"
                              >
                                $
                              </span>
                               echo $EDITOR
                      vim
                      $ git checkout master
                      Switched to branch 'master'
                      Your branch is up-to-date with 'origin/master'.
                      $ git push
                      Everything up-to-date
                      
                            </code>
                          </pre>
                        </div>
                      </div>
                    </div>,
                  }
                }
              >
                <div
                  className="el-scrollbox__content el-scrollbox__content--flexible"
                >
                  <div
                    className="content-reader__layout-h"
                  >
                    <div
                      className="content-reader__lines"
                    >
                      <pre>
                        <code>
                          <div
                            className="content-reader__lines__line-n"
                            key="0"
                          >
                            1
                          </div>
                          <div
                            className="content-reader__lines__line-n"
                            key="1"
                          >
                            2
                          </div>
                          <div
                            className="content-reader__lines__line-n"
                            key="2"
                          >
                            3
                          </div>
                          <div
                            className="content-reader__lines__line-n"
                            key="3"
                          >
                            4
                          </div>
                          <div
                            className="content-reader__lines__line-n"
                            key="4"
                          >
                            5
                          </div>
                          <div
                            className="content-reader__lines__line-n"
                            key="5"
                          >
                            6
                          </div>
                          <div
                            className="content-reader__lines__line-n"
                            key="6"
                          >
                            7
                          </div>
                          <div
                            className="content-reader__lines__line-n"
                            key="7"
                          >
                            8
                          </div>
                        </code>
                      </pre>
                    </div>
                    <div
                      className="content-reader__content"
                      contentEditable={true}
                      onKeyDown={[Function]}
                      role="presentation"
                      spellCheck="false"
                      suppressContentEditableWarning={true}
                    >
                      <pre>
                        <code>
                          $ echo $EDITOR
                      vim
                      $ git checkout master
                      Switched to branch 'master'
                      Your branch is up-to-date with 'origin/master'.
                      $ git push
                      Everything up-to-date
                      
                        </code>
                      </pre>
                    </div>
                  </div>
                </div>
              </ResizeDetector>
            </div>
            <ScrollBar
              handleStyle={Object {}}
              onDrag={[Function]}
              showTrack={false}
              trackStyle={Object {}}
            />
          </div>
        </ResizeDetector>
      </ScrollBox>
    </div>
  </div>
</ContentReader>
`;
